sway-core/src/ir_generation/function.rs

```diff
@@ -2499,17 +2515,39 @@ impl<'eng> FnCompiler<'eng> {
             .add_metadatum(context, whole_block_span_md_idx))
     }
 
+    /// Get the offset, in words, to a particular field in an aggregate type.
+    fn get_offset(
+        &mut self,
+        context: &mut Context,
+        ty: &Type,
+        indices: &[u64],
+    ) -> Result<u64, CompileError> {
+        let mut offset = 0;
+        let mut ty_at_idx = *ty;
+        for index in indices.iter() {
+            let fields = ty_at_idx.get_field_types(context);
+            for (field_idx, field_type) in fields.into_iter().enumerate() {
+                if (field_idx as u64) >= *index {
+                    ty_at_idx = field_type;
+                    break;
+                }
+                offset += ir_type_size_in_bytes(context, &field_type);
+            }
+        }
+        Ok(offset / 8)
+    }
+
```

Comment:
This looks a lot like `aggregate_idcs_to_field_layout()` in `asm_generation`.  The `ir_type_size_in_bytes()` is already stolen from there so you might as well re-use the former.

It's a bit unfortunate that this type-sizing stuff is in IR though, but I understand it's already there, nothing new from this branch.  Ideally we wouldn't be thinking about memory layout until `asm_generation`.   This is a Fuel specific operation though, so assumptions can be made, for now.

Reply:
With regards to target-specific memory layout in IR, LLVM has an analysis pass called [`DataLayout`](https://llvm.org/doxygen/classllvm_1_1DataLayout.html) which is target dependent (relies on the "target triple" in IR) and allows you to query any all kinds of information about how various data types, such as structs, are laid out in memory. This can be quite useful for many optimization passes (even SROA). 

I guess what I'm saying is that querying memory layout in IR (or even in IRGen) is not unheard of.